File: 350.intersection-of-two-arrays-ii.js
/*
 * @lc app=leetcode id=350 lang=javascript
 *
 * [350] Intersection of Two Arrays II
 */

// @lc code=start
/**
 * @param {number[]} nums1
 * @param {number[]} nums2
 * @return {number[]}
 */
var intersect = function (nums1, nums2) {
  const res = [];
  const lookup = {};
  for (const num of nums1) {
    lookup[num] ? ++lookup[num] : (lookup[num] = 1);
  }
  for (const num of nums2) {
    if (lookup[num]) {
      res.push(num);
      --lookup[num];
    }
  }
  return res;
};
// @lc code=end

File: 350.intersection-of-two-arrays-ii.go
/*
 * @lc app=leetcode id=350 lang=golang
 *
 * [350] Intersection of Two Arrays II
 */

package leetcode

// @lc code=start
func intersect(nums1 []int, nums2 []int) []int {
	lookup := map[int]int{}
	res := []int{}
	for i := 0; i < len(nums1); i++ {
		lookup[nums1[i]]++
	}
	for i := 0; i < len(nums2); i++ {
		if lookup[nums2[i]] > 0 {
			lookup[nums2[i]]--
			res = append(res, nums2[i])
		}
	}
	return res
}

// @lc code=end
